Fix issue with sysfs name comparisons.

gonzoleeman opened the pull request:
It turns out that cdev_name_equal() is used
by dlist_find_custom() to compare sysfs
entry names to ones already seen. But it was
comparing using the length of the shortest string
as a maximum, so when it compared, for example,
"eth1" and "eth10", it thought they were the same.
So now just return failure if the strings
aren't the same length, else go ahead and
compare them.

gonzoleeman committed:
Fix issue with sysfs name comparisons.

It turns out that cdev_name_equal() is used
by dlist_find_custom() to compare sysfs
entry names to ones already seen. But it was
comparing using the length of the shortest string
as a maximum, so when it compared, for example,
"eth1" and "eth10", it thought they were the same.
So now just return failure if the strings
aren't the same length, else go ahead and
compare them.

kamalesh-babulal commented on lib/sysfs_class.c:
```diff
@@ -60,13 +60,30 @@ void sysfs_close_class(struct sysfs_class *cls)
+	if (length_a != length_b)
+		return 0;
+
+	if (strncmp(str_a, str_b, SYSFS_NAME_LEN+1) == 0)
```
Length of the string can be less than `SYSFS_NAME_LEN`, the `n` in the `strncmp` should be `length_a` or `length_b`. Throughout the code, the maximum allowed length is `SYSFS_NAME_LEN` which include `\0`, why is `SYSFS_NAME_LEN+1` required here for comparison, what am I missing here?

ematsumiya replied:
Using length_a or length_b for strncmp() is exactly what makes this buggy. See the commit message: it will fail if the system has "eth1" and "eth10" -- i.e. "eth1" will not be shown. One can easily reproduce this by creating a system with >10 NICs and then doing:

```
# systool -c net | grep eth1
  Class Device = "eth10"
  Class Device = "eth11"
```

kamalesh-babulal replied:
Sorry for not being clear, `length_a` and `length_b` are equal after the check `if (length_a != length_b)` but can be lesser than `SYSFS_NAME_LEN`. Using one of them will reflect the comparison better, I don't have any strong opinion on the `n` here.

gonzoleeman replied:
Yes, I was pretty sure that using either length_a or length_b would work, since we know both strings are exactly that length. Or ... is it length_a+1? I always have to figure it out, again. And I knew both strings were less than SYSFS_NAME_LEN+1. :) But I admit I was being lazy. I'll append a commit that changes SYSFS_NAME_LEN with length_a, if that's okay with you (vs pulling this request and submitting a new one).

gonzoleeman replied:
> Using length_a or length_b for strncmp() is exactly what makes this buggy. See the commit message: it will fail if the system has "eth1" and "eth10" -- i.e. "eth1" will not be shown. One can easily reproduce this by creating a system with >10 NICs and then doing:
> 
> ```
> # systool -c net | grep eth1
>   Class Device = "eth10"
>   Class Device = "eth11"
> ```

I think using either length_a or length_b is okay once we establish they are both the same. Since "eth1" and "eth10" have different lengths, they will never get this far.

ematsumiya replied:
Yes, sorry, you're both right. I seem to have ignored the fact that they were being compared right in the lines above.

gonzoleeman committed:
Limit cdev name length comparsion to strlen()+1

In commit 270af678d4241306 cdev_name_equal() was updated
so that "eth1" and "eth10" did not match, but the limit
on the length of the compare was too large, as we know
the length of both strings. This is just a cleanup, and
should cause no functional change.

gonzoleeman commented:
Is this ready for merge now? Or is there still an issue?
